File: Engine/csCore/resource/csJavaFile.cc
#include <csCore/resource/csJavaFile.hh>
#include <csCore/csJava.hh>
#include <csCore/csJavaCalls.hh>

namespace cs
{

#define THIS_JAVA_CLASS "org/cryo/core/resource/CsJavaFile"

static std::string s_name = "";

struct csJavaFilePrivate
{

  csJavaFilePrivate(csJavaFile *file)
  {

    JNIEnv *pEnv = csJava::Get();
    getName = csJavaCallObject<jstring>(pEnv, file, THIS_JAVA_CLASS, "getName", JAVA_STRING);
    open    = csJavaCallBoolean2<jint, jint>(pEnv, file, THIS_JAVA_CLASS, "open", JAVA_INT, JAVA_INT);
    isOpen  = csJavaCallBoolean(pEnv, file, THIS_JAVA_CLASS, "isOpen");
    isEOF   = csJavaCallBoolean(pEnv, file, THIS_JAVA_CLASS, "isEOF");
    seek    = csJavaCallVoid2<jint, jlong>(pEnv, file, THIS_JAVA_CLASS, "seek", JAVA_INT, JAVA_LONG);
    tell    = csJavaCallLong(pEnv, file, THIS_JAVA_CLASS, "tell");
    read    = csJavaCallInt3<jint, jint, jbyteArray>(pEnv,
                                                     file,
                                                     THIS_JAVA_CLASS,
                                                     "read",
                                                     JAVA_INT,
                                                     JAVA_INT,
                                                     JAVA_BYTE_ARRAY);
    close   = csJavaCallVoid(pEnv, file, THIS_JAVA_CLASS, "close");
  }

  csJavaCallObject<jstring>              getName;
  csJavaCallBoolean2<jint, jint>         open;
  csJavaCallBoolean                      isOpen;
  csJavaCallBoolean                      isEOF;
  csJavaCallVoid2<jint, jlong>           seek;
  csJavaCallLong                         tell;
  csJavaCallInt3<jint, jint, jbyteArray> read;
  csJavaCallVoid                         close;

};


csJavaFile::csJavaFile()
    : m_priv(new csJavaFilePrivate(this))
{

}

csJavaFile::~csJavaFile()
{
  delete m_priv;
  m_priv = nullptr;
}


const std::string &csJavaFile::GetName() const
{
  jstring res = m_priv->getName.call(csJava::Get(), 0);
  if (res)
  {
    m_name = csJava::Get()->GetStringUTFChars(res, 0);
  }
  return m_name;
}
bool csJavaFile::Open(eAccessMode accessMode, eOpenMode openMode)
{
  return m_priv->open.call(csJava::Get(), accessMode, openMode, false);
}
bool csJavaFile::IsOpen() const
{
  return m_priv->isOpen.call(csJava::Get(), false);
}

bool csJavaFile::IsEOF() const
{
  return m_priv->isEOF.call(csJava::Get(), false);

}
void csJavaFile::Seek(eSeekMode mode, long offset)
{
  m_priv->seek.call(csJava::Get(), mode, offset);
}


long csJavaFile::Tell() const
{
  return (long) m_priv->tell.call(csJava::Get(), 0);
}

Size csJavaFile::Read(Size elementSize, Size elementCount, void *buffer)
{
  size_t     size   = elementSize * elementCount;
  JNIEnv     *pEnv  = csJava::Get();
  jbyteArray pArray = pEnv->NewByteArray((jsize) size);



  jint readBytes = m_priv->read.call(pEnv, elementSize, elementCount, pArray, 0);

  jbyte *bytes = pEnv->GetByteArrayElements(pArray, 0);
  memcpy(buffer, bytes, readBytes);
  pEnv->ReleaseByteArrayElements(pArray, bytes, 0);


  pEnv->DeleteLocalRef(pArray);

  return readBytes;
}
Size csJavaFile::Write(Size elementSize, Size elementCount, const void *buffer)
{
  return 0;
}
void csJavaFile::Close()
{
  m_priv->close.call(csJava::Get());
}

}

File: JCryoStorm/CsCore/src/main/java/org/cryo/core/resource/CsJavaFile.java
package org.cryo.core.resource;


import org.cryo.core.CsClass;
import org.cryo.core.CsObject;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.nio.ByteBuffer;

import static org.cryo.core.resource.CsJavaFileNative.*;

@CsClass(CS_CLASS_NAME)
public class CsJavaFile extends CsObject implements IFile {

    private final String name;

    private final byte[] data;

    private final InputStream inputStream;

    private int ptr;

    public CsJavaFile(String name, InputStream inputStream) {
        super();
        this.name = name;
        this.inputStream = inputStream;
        this.data = readAll(inputStream);
        this.ptr = 0;
    }

    @Override
    public String getName() {
        return this.name;
    }

    public boolean open(int accessMode, int openMode) {
        return open(EAccessMode.values()[accessMode],
                    EOpenMode.values()[openMode]);
    }

    @Override
    public boolean open(EAccessMode accessMode, EOpenMode openMode) {
        return this.data != null && this.inputStream != null;
    }

    @Override
    public boolean isOpen() {
        return this.data != null && this.inputStream != null;
    }

    @Override
    public boolean isEOF() {
        return this.ptr >= this.data.length;
    }

    public void seek(int seekMode, long offset) {
        seek(ESeekMode.values()[seekMode], offset);
    }

    @Override
    public void seek(ESeekMode seekMode, long offset) {
        switch (seekMode) {
            case SET:
                this.ptr = (int) offset;
                break;
            case CUR:
                this.ptr += offset;
                break;
            case END:
                this.ptr = this.data.length - (int) offset;
                break;
        }
    }

    @Override
    public long tell() {
        return this.ptr;
    }

    @Override
    public int read(int elementSize, int elementCount, byte[] byteBuffer) {
        int numBytes = elementCount * elementSize;
        int remaining = this.data.length - this.ptr;
        int bytesRead = Math.min(numBytes, remaining);


        for (int i = 0; i < bytesRead; i++) {
            byteBuffer[i] = this.data[this.ptr++];
        }

        return bytesRead;
    }

    @Override
    public int write(int elementSize, int elementCount, ByteBuffer byteBuffer) {
        return 0;
    }

    @Override
    public void close() {
        try {
            this.inputStream.close();
        }
        catch (Exception e) {
            //
        }
    }


    private static byte[] readAll(InputStream inputStream) {
        try {
            // Create a ByteArrayOutputStream
            ByteArrayOutputStream buffer = new ByteArrayOutputStream();

            // Create a byte array for the buffer
            int nRead;
            byte[] data = new byte[1024];

            // Read the bytes from the input stream
            while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
                buffer.write(data, 0, nRead);
            }

            // Get the byte array
            return buffer.toByteArray();
        }
        catch (Exception e) {
            return null;
        }
    }
}
